src/main/java/com/minecolonies/entity/ai/citizen/lumberjack/Tree.java

```diff
@@ -193,13 +208,66 @@ public static boolean checkTree(@NotNull IBlockAccess world, BlockPos pos)
             return false;
         }
 
+        BlockPos[] baseAndTOp = getBottomAndTopLog(world, pos, new LinkedList<>(), null, null);
+
+        if(baseAndTOp == null || baseAndTOp.length < 2)
+        {
+            return false;
+        }
+
         //Get base log, should already be base log
-        BlockPos basePos = getBaseLog(world, pos);
+        BlockPos basePos = baseAndTOp[0];
 
         //Make sure tree is on solid ground and tree is not build above cobblestone
         return world.getBlockState(basePos.down()).getMaterial().isSolid()
                  && world.getBlockState(basePos.down()).getBlock() != Blocks.COBBLESTONE
-                 && hasEnoughLeaves(world, pos);
+                 && hasEnoughLeaves(world, baseAndTOp[1]);
+    }
+
+    /**
+     * Adds a log and searches for further logs(Breadth first search)
+     *
+     * @param world The world the log is in
+     * @param log   the log to add
+     */
+    private static BlockPos[] getBottomAndTopLog(@NotNull IBlockAccess world, @NotNull BlockPos log, @NotNull LinkedList<BlockPos> woodenBlocks, BlockPos bottomLog,
+            BlockPos topLog)
+    {
+        BlockPos bottom = bottomLog == null ? log  : bottomLog;
+        BlockPos top = topLog == null ? log : topLog;
+
+        if(woodenBlocks.size() >= MAX_TREE_SIZE)
+        {
+            return new BlockPos[]{bottom, top};
+        }
+
+        if(log.getY() < bottom.getY())
+        {
+            bottom = log;
+        }
+
+        if(log.getY() > top.getY())
+        {
+            top = log;
+        }
+
+        woodenBlocks.add(log);
+        for (int y = -1; y <= 1; y++)
+        {
+            for (int x = -1; x <= 1; x++)
+            {
+                for (int z = -1; z <= 1; z++)
+                {
+                    BlockPos temp = log.add(x, y, z);
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'temp' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)